feat(analyzer): add rules regarding model name suffixes

archerzz opened the pull request:
- add four rules checking the suffixes of model names
- upgrade Roslyn to latest version since new rules require some API
- minor refactor existing codes
- add test cases for the new rules

resolve #6905

archerzz commented:
preview pr on `azure-sdk-for-net`: https://github.com/Azure/azure-sdk-for-net/pull/38560

@m-nash ^^

christothes commented on src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ModelName/SuffixAnalyzerBase.cs:
```diff
@@ -0,0 +1,79 @@
+        }
+
+        protected virtual bool ShouldSkip(INamedTypeSymbol symbol, SymbolAnalysisContext context) => false;
+        protected virtual Regex SuffixRegex => throw new NotImplementedException();
```
Since these regexes are just doing the equivalent of "EndsWith", it would be more performant to make this something like below 

```suggestion
        protected abstract string[] InvalidSuffixes;
```

Then the implementation of `AnalyzeSymbol` changes to iterate through InvalidSuffixes to see if `typeSymbol.Name` ends with any of them.

I did a quick perf test just to verify that this is indeed faster, even with checking the worst case of the general suffixes.

|   Method |       Mean |     Error |   StdDev | Ratio | Allocated | Alloc Ratio |
|--------- |-----------:|----------:|---------:|------:|----------:|------------:|
| EndsWith |   193.9 ns |   3.76 ns |  4.48 ns |  0.02 |         - |          NA |
|    RegEx | 8,544.4 ns | 100.82 ns | 94.30 ns |  1.00 |         - |          NA |

archerzz replied:
replaced with strings

archerzz commented on src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ModelName/SuffixAnalyzerBase.cs:
```diff
@@ -0,0 +1,79 @@
+            }
+        }
+
+        protected virtual bool ShouldSkip(INamedTypeSymbol symbol, SymbolAnalysisContext context) => false;
+        protected virtual Regex SuffixRegex => throw new NotImplementedException();
+        protected virtual Diagnostic GetDiagnostic(INamedTypeSymbol typeSymbol, string suffix, SymbolAnalysisContext context) => throw new NotImplementedException();
```
fixed, thanks

archerzz commented on src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/ModelName/AZC0030Tests.cs:
```diff
@@ -0,0 +1,114 @@
+    public class AZC0030Tests
+    {
+        [Fact]
+        public async Task ClassNotUnderModelsNamespaceIsNotChecked()
```
added check for serialization and deserialization methods

archerzz commented on src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ModelName/SuffixAnalyzerBase.cs:
```diff
@@ -0,0 +1,79 @@
+    {
+        protected static readonly string Title = "Improper model name suffix";
+        protected static readonly string Description = "Suffix is not recommended. Consider to remove or modify it.";
+        protected static readonly string GeneralRenamingMessageFormat = "Model name '{0}' ends with '{1}'. Suggest to rename it to an appropriate name.";
```
I could not find the official guidelines. We do have a [OneNote page](https://microsoft.sharepoint.com/:o:/t/AzureDeveloperExperience/En9PCC8lAB1DpXdlD70PfToBzewQyWSM79-9iJR5G60Vkw?e=mbaUtp) on that. @m-nash Shall we update the [.NET Azure SDK Design Guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html)

archerzz replied:
Or we can do something like [StyleCop](https://github.com/DotNetAnalyzers/StyleCopAnalyzers).
- For each diagnostic ID, we create a page, like https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA0001.md
- In each page, we provide explanation of the diagnostic, and provide mitigation solution.

christothes commented on src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/INameSpaceSymbolExtensions.cs:
```diff
@@ -1,20 +1,28 @@
+        public static string GetFullNamespaceName(this INamespaceSymbol namespaceSymbol)
         {
-            if (namespaceSymbo is { ContainingNamespace: not null and { IsGlobalNamespace: false} })
+            return string.Join(".", namespaceSymbol.GetAllNamespaces());
```
Perhaps we should re-think this implementation a bit. We are building a list that will end up in a `String.Join`, should we consider a StringBuilder? We can also avoid the list `Reverse` of the list by appending at position 0 of the builder.

Also, given that the [consumer](https://github.com/Azure/azure-sdk-tools/pull/6916/files/2e313f401eeb8adc5f3d4c7c1ec18eaace8af0be..9abf0865f130aee50f3827567e7b5e1845a962e4#diff-b7bcf510670ac60c111caaae450c5b661b6e32cd4f75f8f55f42e2b8c1416d71R39) of this turns the string back to a stringBuilder, perhaps this should just return a string builder?

archerzz commented:
Changes since last review:
- replace `List` and `Join` with `StringBuilder`
- adopt `ArrayPool` to collect namespaces
- merge latest `main` branch
- extract analyzer descriptors to align with other analyzers
